### DISCUSSION TITLE
custom formatters and dates

### DISCUSSION THREAD

**Discussion by billcla** [anvilistas/tabulator#141] [Q&A] [answered]
Hello,

I've been playing around with the Tabulator for one of my apps and its fantastic but i'm having a couple of issues that I cant seem to over come.
I have a column which is set up as a date:
`{"title":"To", 
"field":"sc_ferm_to", 
"width":100, "resizable": True, 
"visible" : False, 
"sorter": "date",
 "formatter": custom_formatter_date, 
"formatter_params": {"format": "%d %b %Y"}},`

The custom formatter is:
`def custom_formatter_date(cell, **params):`
`      value2 = cell.get_value()`
`      if value2 is not None:`
`        cell.get_element().style.backgroundColor = "yellow"`
`       print(value2)`
`       return value2`

Whilst the code gives the cell a background colour as expected, value2 doesn't seem to be returned (although printing it gives the detail in the console). If I apply this custom formatter to any other cell which is just text rather than a date it works fine, the cell is given the correct background colour and the text appears in the cell.
For example:
![image](https://github.com/anvilistas/tabulator/assets/53837913/f453e7a7-3766-426d-85bd-abcc2756aeba)

I know dates are a little trickier in tabulator, am I missing something?

**Comment by s-cork** [comment]
Could you create a minimal example clone link and I can take a look.

**Reply by billcla** [reply]
Hi Stu,

Much appreciated! i have built a mini clone. you'll see in the server module I am returning a list of dicts to the client because in my main app i am doing a bit of processing of linked tables to return additional information (if the linked row contains the user return True) so have emulated the same in the clone, being the records are returned as a list of dicts (in case that is the problem, cant see a reason why though). 

I get the same result in that the cell for the "to" date is formatted with the colour (or not) but the date itself is not returned, however the custom_formatter_date is able to print the cell_val just before the return call.

https://anvil.works/build#clone:IZGXOW4Q7JFC7BZG=PNGZ6WKV3S3FYNEZ7CBXAAXT|null

any ideas?

**Comment by s-cork** [comment, marked as answer]
a cell formatter expects the return value to be a string
you're returning a datetime object

so something like

```python
return cell_val.isoformat()
```

will sort it for you

**Reply by billcla** [reply]
What a fool... my apologies I completely overlooked that fact!! Many thanks!
